vpr/src/route/connection_box_lookahead_map.cpp

```diff
@@ -92,78 +120,65 @@ class CostMap {
-                      const t_routing_cost_map& cost_map,
-                      e_representative_entry_method method) {
-        const auto& device_ctx = g_vpr_ctx.device();
-        for (size_t box_id = 0;
-             box_id < device_ctx.connection_boxes.num_connection_box_types();
-             ++box_id) {
-            set_cost_map(from_seg_index, ConnectionBoxId(box_id), cost_map, method);
-        }
-    }
-
-    void set_cost_map(int from_seg_index, ConnectionBoxId box_id, const t_routing_cost_map& cost_map, e_representative_entry_method method) {
-        VTR_ASSERT(from_seg_index >= 0 && from_seg_index < (ssize_t)offset_.size());
+    // Set cost map from sparse hash table of data.
+    void set_cost_map(const routing_data_t& data, e_representative_entry_method method) {
+        vtr::NdMatrix<t_bb, 2> bounds({cost_map_.dim_size(0), cost_map_.dim_size(1)},
+                                      {0, 0, 0, 0});
 
         // Find coordinate offset for this segment.
-        int min_dx = 0;
-        int min_dy = 0;
-        int max_dx = 0;
-        int max_dy = 0;
-        for (const auto& entry : cost_map) {
-            if (std::get<3>(entry) != box_id) {
-                continue;
-            }
-            min_dx = std::min(std::get<1>(entry).first, min_dx);
-            min_dy = std::min(std::get<1>(entry).second, min_dy);
-
-            max_dx = std::max(std::get<1>(entry).first, max_dx);
-            max_dy = std::max(std::get<1>(entry).second, max_dy);
-        }
+        for (const auto& entry : data) {
+            t_bb& bound = bounds[std::get<0>(entry.first)][size_t(std::get<1>(entry.first))];
 
-        offset_[from_seg_index][size_t(box_id)].first = min_dx;
-        offset_[from_seg_index][size_t(box_id)].second = min_dy;
-        size_t dim_x = max_dx - min_dx + 1;
-        size_t dim_y = max_dy - min_dy + 1;
+            bound.xmin = std::min(std::get<2>(entry.first), bound.xmin);
+            bound.xmax = std::max(std::get<2>(entry.first), bound.xmax);
 
-        vtr::NdMatrix<Expansion_Cost_Entry, 2> expansion_cost_map(
-            {dim_x, dim_y});
+            bound.ymin = std::min(std::get<3>(entry.first), bound.ymin);
+            bound.ymax = std::max(std::get<3>(entry.first), bound.ymax);
+        }
 
-        for (const auto& entry : cost_map) {
-            if (std::get<3>(entry) != box_id) {
-                continue;
+        // Allocate cost matricies based on bounds.
+        for (size_t from_seg_index = 0; from_seg_index < cost_map_.dim_size(0); ++from_seg_index) {
+            for (size_t box_id = 0; box_id < cost_map_.dim_size(1); ++box_id) {
+                const t_bb& bound = bounds[from_seg_index][box_id];
+                offset_[from_seg_index][box_id].first = bound.xmin;
+                offset_[from_seg_index][box_id].second = bound.ymin;
+                size_t dim_x = bound.xmax - bound.xmin + 1;
+                size_t dim_y = bound.ymax - bound.ymin + 1;
```

Comment:
Should be `bound.witfh()` & `bound.height()`.

Reply:
Also, `vtr::Rect` already has these implememted. I'm not sure why there's also `t_bb`.

Reply:
In this specific case, changing from this computation to bound.width() is actually an off by one bug, because vtr::Rect is inclusive on the low bound, but exclusive on the high bound.  t_bb doesn't define a width operator, but its usage is inclusive on xmin and xmax.